File: src/fta_router/dataset.py
"""Load and validate behavioural routing JSONL datasets."""

from __future__ import annotations

import json
from collections import Counter
from collections.abc import Iterable, Iterator, Mapping, Sequence
from pathlib import Path
from typing import Any

from .schema import RoutingExample, SchemaError, validate_example


class DatasetError(ValueError):
    """Raised when a JSONL dataset cannot be loaded or fails validation."""

    def __init__(self, errors: list[str]) -> None:
        self.errors = list(errors)
        super().__init__("; ".join(self.errors) if self.errors else "invalid dataset")


def iter_jsonl(path: str | Path) -> Iterator[tuple[int, dict[str, Any]]]:
    """Yield ``(line_no, row)`` for each non-empty JSONL record."""
    path = Path(path)
    with path.open(encoding="utf-8") as handle:
        for line_no, line in enumerate(handle, start=1):
            line = line.strip()
            if not line:
                continue
            try:
                row = json.loads(line)
            except json.JSONDecodeError as exc:
                raise DatasetError([f"{path}:{line_no}: invalid JSON: {exc}"]) from exc
            if not isinstance(row, dict):
                raise DatasetError(
                    [f"{path}:{line_no}: JSON object required, got {type(row).__name__}"]
                )
            yield line_no, row


def load_jsonl(path: str | Path) -> list[dict[str, Any]]:
    """Load raw dictionaries from a JSONL file (skips blank lines)."""
    return [row for _, row in iter_jsonl(path)]


def load_examples(path: str | Path, *, validate: bool = True) -> list[RoutingExample]:
    """Load ``RoutingExample`` records from JSONL.

    When *validate* is True (default), schema and uniqueness errors raise
    ``DatasetError``.
    """
    if validate:
        errors = validate_file(path)
        if errors:
            raise DatasetError(errors)
    return [RoutingExample.from_dict(row, validate=False) for row in load_jsonl(path)]


def write_jsonl(path: str | Path, rows: Sequence[Mapping[str, Any]]) -> None:
    """Write mapping rows as JSONL, creating parent directories as needed."""
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8") as handle:
        for row in rows:
            handle.write(json.dumps(dict(row), ensure_ascii=True) + "\n")


def validate_file(path: str | Path) -> list[str]:
    """Validate all rows in a JSONL file; return error messages (empty if OK)."""
    path = Path(path)
    errors: list[str] = []
    seen_ids: set[str] = set()
    try:
        stream = path.open(encoding="utf-8")
    except OSError as exc:
        return [f"cannot read {path}: {exc}"]

    with stream:
        for line_no, line in enumerate(stream, start=1):
            line = line.strip()
            if not line:
                continue
            try:
                row = json.loads(line)
            except json.JSONDecodeError as exc:
                errors.append(f"{path.name}:{line_no}: invalid JSON: {exc}")
                continue
            if not isinstance(row, dict):
                errors.append(f"{path.name}:{line_no}: JSON object required")
                continue
            errors.extend(validate_example(row, line_no=line_no))
            rid = row.get("id")
            if isinstance(rid, str) and rid:
                if rid in seen_ids:
                    errors.append(f"{path.name}:{line_no}: duplicate id '{rid}'")
                seen_ids.add(rid)
    return errors


def validate_rows(rows: Iterable[Mapping[str, Any]]) -> list[str]:
    """Validate in-memory rows (including duplicate ids)."""
    errors: list[str] = []
    seen_ids: set[str] = set()
    for index, row in enumerate(rows, start=1):
        errors.extend(validate_example(row, line_no=index))
        rid = row.get("id")
        if isinstance(rid, str) and rid:
            if rid in seen_ids:
                errors.append(f"line {index}: duplicate id '{rid}'")
            seen_ids.add(rid)
    return errors


def label_distribution(
    rows: Iterable[Mapping[str, Any]],
    field: str = "primary_action",
) -> dict[str, int]:
    """Count values of *field* across rows."""
    return dict(Counter(str(row[field]) for row in rows))


def summarise(path: str | Path) -> dict[str, Any]:
    """Return counts for a JSONL routing file."""
    rows = load_jsonl(path)
    return {
        "path": str(path),
        "n": len(rows),
        "primary_action": label_distribution(rows, "primary_action"),
        "domain": label_distribution(rows, "domain"),
        "reasoning_level": label_distribution(rows, "reasoning_level"),
    }


def examples_to_dicts(examples: Sequence[RoutingExample]) -> list[dict[str, Any]]:
    """Serialize examples back to plain dictionaries."""
    return [example.to_dict() for example in examples]


# Re-export for callers that want a single import surface.
__all__ = [
    "DatasetError",
    "SchemaError",
    "examples_to_dicts",
    "iter_jsonl",
    "label_distribution",
    "load_examples",
    "load_jsonl",
    "summarise",
    "validate_file",
    "validate_rows",
    "write_jsonl",
]

File: javascript/src/dataset.ts
/** Load and validate behavioural routing JSONL datasets. */

import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";

import { RoutingExample, SchemaError, validateExample, type RoutingRecord } from "./schema.ts";

export class DatasetError extends Error {
  readonly errors: string[];

  constructor(errors: string[]) {
    const list = [...errors];
    super(list.length ? list.join("; ") : "invalid dataset");
    this.name = "DatasetError";
    this.errors = list;
  }
}

export function iterJsonl(filePath: string): Array<[number, Record<string, unknown>]> {
  const rows: Array<[number, Record<string, unknown>]> = [];
  let text: string;
  try {
    text = readFileSync(filePath, "utf8");
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new DatasetError([`cannot read ${filePath}: ${message}`]);
  }
  const lines = text.split(/\r?\n/);
  for (let i = 0; i < lines.length; i += 1) {
    const lineNo = i + 1;
    const line = lines[i].trim();
    if (!line) {
      continue;
    }
    let row: unknown;
    try {
      row = JSON.parse(line);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      throw new DatasetError([`${filePath}:${lineNo}: invalid JSON: ${message}`]);
    }
    if (row === null || typeof row !== "object" || Array.isArray(row)) {
      throw new DatasetError([
        `${filePath}:${lineNo}: JSON object required, got ${Array.isArray(row) ? "array" : typeof row}`,
      ]);
    }
    rows.push([lineNo, row as Record<string, unknown>]);
  }
  return rows;
}

export function loadJsonl(filePath: string): Record<string, unknown>[] {
  return iterJsonl(filePath).map(([, row]) => row);
}

export function loadExamples(filePath: string, options: { validate?: boolean } = {}): RoutingExample[] {
  const validate = options.validate ?? true;
  if (validate) {
    const errors = validateFile(filePath);
    if (errors.length) {
      throw new DatasetError(errors);
    }
  }
  return loadJsonl(filePath).map((row) => RoutingExample.fromDict(row, { validate: false }));
}

export function writeJsonl(filePath: string, rows: Array<Record<string, unknown> | RoutingRecord>): void {
  mkdirSync(path.dirname(filePath), { recursive: true });
  const body = rows.map((row) => JSON.stringify(row)).join("\n") + (rows.length ? "\n" : "");
  writeFileSync(filePath, body, "utf8");
}

export function validateFile(filePath: string): string[] {
  const errors: string[] = [];
  const seenIds = new Set<string>();
  let text: string;
  try {
    text = readFileSync(filePath, "utf8");
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return [`cannot read ${filePath}: ${message}`];
  }
  const name = path.basename(filePath);
  const lines = text.split(/\r?\n/);
  for (let i = 0; i < lines.length; i += 1) {
    const lineNo = i + 1;
    const line = lines[i].trim();
    if (!line) {
      continue;
    }
    let row: unknown;
    try {
      row = JSON.parse(line);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      errors.push(`${name}:${lineNo}: invalid JSON: ${message}`);
      continue;
    }
    if (row === null || typeof row !== "object" || Array.isArray(row)) {
      errors.push(`${name}:${lineNo}: JSON object required`);
      continue;
    }
    const record = row as Record<string, unknown>;
    errors.push(...validateExample(record, { lineNo }));
    const rid = record.id;
    if (typeof rid === "string" && rid) {
      if (seenIds.has(rid)) {
        errors.push(`${name}:${lineNo}: duplicate id '${rid}'`);
      }
      seenIds.add(rid);
    }
  }
  return errors;
}

export function validateRows(rows: Array<Record<string, unknown>>): string[] {
  const errors: string[] = [];
  const seenIds = new Set<string>();
  rows.forEach((row, index) => {
    const lineNo = index + 1;
    errors.push(...validateExample(row, { lineNo }));
    const rid = row.id;
    if (typeof rid === "string" && rid) {
      if (seenIds.has(rid)) {
        errors.push(`line ${lineNo}: duplicate id '${rid}'`);
      }
      seenIds.add(rid);
    }
  });
  return errors;
}

export function labelDistribution(
  rows: Array<Record<string, unknown>>,
  field = "primary_action",
): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const row of rows) {
    const key = String(row[field]);
    counts[key] = (counts[key] ?? 0) + 1;
  }
  return counts;
}

export function summarise(filePath: string): {
  path: string;
  n: number;
  primary_action: Record<string, number>;
  domain: Record<string, number>;
  reasoning_level: Record<string, number>;
} {
  const rows = loadJsonl(filePath);
  return {
    path: filePath,
    n: rows.length,
    primary_action: labelDistribution(rows, "primary_action"),
    domain: labelDistribution(rows, "domain"),
    reasoning_level: labelDistribution(rows, "reasoning_level"),
  };
}

export const summarize = summarise;

export function examplesToDicts(examples: RoutingExample[]): RoutingRecord[] {
  return examples.map((example) => example.toDict());
}

export { SchemaError };
